Resurrect Bunny Dance

colemanw opened the pull request:
Overview
----------------------------------------
This resolves https://lab.civicrm.org/dev/core/-/issues/3722 with a new approach to #23955 to prevent an infinite loop.

Before
-------------
In unit tests, `BunnyDance.mgd.php` was partially commented out, because uncommenting it gives:

```
Fatal error: Allowed memory size of 2147483648 bytes exhausted (tried to al  
  locate 20480 bytes) in /home/homer/buildkit/build/build-0/web/sites/all/mod  
  ules/civicrm/vendor/pear/pear-core-minimal/src/PEAR.php on line 898          
  Remove /home/homer/buildkit/build/build-0/web/sites/default/files/civicrm/e  
  xt/example-mixin                                                             
  Create /home/homer/buildkit/build/build-0/web/sites/default/files/civicrm/e  
  xt/example-mixin for scan-classes@1                                          
  Test scan-classes@1     
```

After
---------
Fatal infinite loop fixed.
Test fixed.

Technical Details
---------------
The problem #23955 wanted to solve was an infinite loop being caused by `Managed->reconcile()` calling itself.
This fixes that problem. And I've confirmed that existing tests do cover the functionality:
- `CRM_Case_BAO_CaseTypeForkTest.testManagerContact`
- `api_v3_CaseTypeTest.testCaseTypeDelete_InUse`

The `ManagedCaseTypeTest` turned out to be broken due to double-caching of the case types pseudoconstant, which was persisting beyond cache flushes. Fixed by removing the double-caching.

Also, while I was looking at it I optimized the `CRM_Case_ManagedEntities` reconcile and dropped some unused variables.

herbdool committed:
Update BunnyDance.mgd.php

colemanw committed:
CiviCase - Prevent recursion when reconciling managed activity/relationship types & fix test

The createManagedRelationshipTypes and createManagedActivityTypes functions always use 'civicrm'
as the module name, so this reconciliation can target just that module to prevent recursion.

The ManagedCaseTypeTest was failing due to caching/timing issues.

Try this...

eileenmcnaughton commented on mixin/mgd-php@1/example/CRM/BunnyDance.mgd.php:
```diff
@@ -14,41 +14,40 @@
+            ['name' => 'Change Case Type'],
+            ['name' => 'Change Case Status'],
+            ['name' => 'Change Case Start Date'],
+            ['name' => 'Link Cases'],
```
This one should be 'Eat furniture'

colemanw replied:
Ah, I had a gnawing feeling that was the problem.

eileenmcnaughton replied:
lol

colemanw committed:
CiviCase - Optimize managed reconcile

colemanw commented:
retest this please

demeritcowboy commented:
I'm not familiar with the bunny dance but the [hamster dance](https://www.youtube.com/watch?v=WEH2fk0ONag) was all the rage back when `<blink>` was cool.
Anyway nice job wrestling the bunny.
And I'll note xml files still work with this.

colemanw committed:
CRM_Case_ManagedEntities - Fix bugs caused by double-caching

The list of case types was being cached twice, once by CRM_Core_Pseudoconstant
and yet again in CRM_Case_XMLRepository::$allCaseTypes.
This made getting a fresh list of case types unnecessarily difficult.

totten commented:
Sweet. Not only does the bunny dance, but it also finds deep-gnarly bugs.

If the mixin-tests are passing, then I believe.

Nicely done @colemanw
